### PR TITLE
Support metadata encoding

### PR DESCRIPTION
**Pull request by Isotr0py**
- Support encode with metadata
- Add `encode_with_box` to encode image with metadata

By default the encoder assumes no metadata boxes will be added according to [libjxl's API](https://libjxl.readthedocs.io/en/latest/api_encoder.html#_CPPv418JxlEncoderUseBoxesP10JxlEncoder). So metadata like EXIF data is missing when creating `jxl` from png image etc.

An example code to encode with EXIF data:
```rust
use std::io::Write;

use jpegxl_rs::encode::{EncoderResult, BoxTypes};
use jpegxl_rs::encoder_builder;
use image::io::Reader as ImageReader;

fn main() {
    let sample = ImageReader::open("sample.png").unwrap().decode().unwrap().to_rgb8();
    let mut encoder = encoder_builder().build().unwrap();

    let file = std::fs::File::open("sample.png").unwrap();
    let mut bufreader = std::io::BufReader::new(&file);
    let exifreader = exif::Reader::new();
    let exif = exifreader.read_from_container(&mut bufreader).unwrap();

    let buffer: EncoderResult<f32> = encoder.encode_with_box(&sample, sample.width(), sample.height(), exif.buf(), BoxTypes::Exif).unwrap();
    
    let mut file = std::fs::File::create("sample.jxl").unwrap();
    let _ = file.write(&buffer.data);
}

```

### CODE REVIEW

File: jpegxl-rs/src/encode.rs [hunk 1]
```diff
@@ -92,6 +92,17 @@ impl From<ColorEncoding> for JxlColorEncoding {
     }
 }
 
+/// Encoding box types
+#[derive(Debug, Clone, Copy)]
+pub enum BoxTypes {
```
Comment (inflation) [inline jpegxl-rs/src/encode.rs:97]: How about a `Metadata<'data>` with `Exif<&'data u8>` and etc. ?

File: jpegxl-rs/src/encode.rs [hunk 2]
```diff
@@ -525,6 +555,26 @@ impl<'prl, 'mm> JxlEncoder<'prl, 'mm> {
...
+        &mut self,
+        data: &[T],
+        width: u32,
+        height: u32,
+        box_contents: &[u8],
+        box_types: BoxTypes,
+    ) -> Result<EncoderResult<U>, EncodeError> {
+        self.setup_encoder(width, height, U::bits_per_sample(), self.has_alpha)?;
+        self.add_frame(&EncoderFrame::new(data))?;
+        self.add_box(box_contents, box_types)?;
```
Comment (inflation) [inline jpegxl-rs/src/encode.rs:574]: You need to call `JxlEncoderCloseBoxes` before starting to encode the pixels.

File: jpegxl-rs/src/encode.rs [hunk 3]
```diff
@@ -387,6 +398,25 @@ impl JxlEncoder<'_, '_> {
...
+            BoxTypes::Jumb => [106, 117, 109, 98],
+        };
+        self.check_enc_status(unsafe { JxlEncoderUseBoxes(self.enc) })?;
+        self.check_enc_status(unsafe {
+            JxlEncoderAddBox(
+                self.enc,
+                &mut box_type,
+                box_contents.as_ptr().cast(),
+                box_contents.len(),
+                false.into(),
```
Comment (inflation) [inline jpegxl-rs/src/encode.rs:415]: We should expose compression capability if `libjxl` supports it.

**Review by inflation — COMMENTED** [review]
Thanks for the PR!

One thing I'd like to point out is that `jxl` supports multiple metadata boxes. We should probably follow the API design for `EncoderFrame`.

File: jpegxl-rs/src/encode.rs (hunk 1 above)
Comment (Isotr0py) [reply jpegxl-rs/src/encode.rs]: Got it! We can use `Metadata` to collect multiple metadata contents and other metadata configuration.

But maybe we can remain this to use in box type matching with `c_char` when adding box.

**Pull request closed without merging**